consensus/polybft/stake_manager.go

```diff
@@ -0,0 +1,368 @@
+	}
+
+	return s.state.StakeStore.insertTransferEvents(epoch, events)
+}
+
+// UpdateValidatorSet returns an updated validator set
+// based on stake change (transfer) events from ValidatorSet contract
+func (s *stakeManager) UpdateValidatorSet(epoch uint64, currentValidatorSet AccountSet) (*ValidatorSetDelta, error) {
+	s.logger.Info("Calculating validators set update...", "epoch", epoch)
+
+	transferEvents, err := s.state.StakeStore.getTransferEvents(epoch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get transfer events for epoch: %d. Error: %w", epoch, err)
+	}
+
+	if len(transferEvents) == 0 {
+		s.logger.Info("Calculating validators set finished. No transfer events for given epoch.",
+			"epoch", epoch)
+
+		return &ValidatorSetDelta{}, nil
+	}
+
+	// stake counter holds sorted stakes by current and (possible) new validators
+	// he will add to map current stake (voting power) of the current validators
+	// on object instantiation
+	stakeCounter := newStakeCounter(currentValidatorSet)
+
+	// update the stake counter with stake changes from transfer events
+	for _, event := range transferEvents {
+		if event.IsStake() {
+			// then this amount was minted To validator address
+			stakeCounter.addStake(event.To, event.Value)
+		} else if event.IsUnstake() {
+			// then this amount was burned From validator address
+			stakeCounter.removeStake(event.From, event.Value)
+		} else {
+			// this should not happen, but lets log it if it does
+			s.logger.Debug("Found a transfer event that represents neither stake nor unstake")
+		}
+	}
+
+	// sort validators by stake since we update the validator set
+	// based on highest stakes
+	stakeCounter.sortByStake()
+
+	removedBitmap := &bitmap.Bitmap{}
+	updatedValidators := AccountSet{}
+	addedValidators := AccountSet{}
+
+	// first deal with existing validators
+	nonChangedValidators := 0
+
+	for i, a := range currentValidatorSet.Copy() {
+		stakeInfo := stakeCounter.getStake(a.Address)
+
+		if stakeInfo.pos > s.maxValidatorSetSize-1 {
+			// validator is not in the maximum validator set we support based on its stake
+			// so we will remove it
+			removedBitmap.Set(uint64(i))
+		} else {
+			if stakeInfo.stake.Cmp(bigZero) == 0 {
+				// validator un-staked all, remove it from validator set
+				removedBitmap.Set(uint64(i))
+			} else if stakeInfo.stake.Cmp(a.VotingPower) != 0 {
+				// validator updated its stake so put it in the updated validators list
+				a.VotingPower = new(big.Int).Set(stakeInfo.stake)
+				updatedValidators = append(updatedValidators, a)
+			} else {
+				// he did not change stake, but he's left in validator set
+				nonChangedValidators++
+			}
+		}
+	}
+
+	// add new validators until we reach the max validator set size
+	for _, si := range stakeCounter.iterateThroughNewValidators() {
+		if si.pos > s.maxValidatorSetSize-1 {
+			// new validator doesn't have enough stake to be in validator set
+			continue
```

Comment:
can we at least put log here?

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/1431/commits/ce5d87c88920b208f4834629f6fe9c135d997e4f